Port Reuse: Would love some help - unit tests only pass most of the time.

markmandel opened the pull request:
Draft pull request, in hopes that someone else will see something that I am missing. I'd love an extra pair of eyes, because I'm at a bit of a loss here. 😞

Please excuse the mess, work in progress code.

## Context

I believe I've implemented the basics of #410, but on my unit tests, it's passing ~90% of the time, but 10% of the time, Quilkin isn't erroring when it says it's receiving UDP packets, but when the test sends them, nothing happens and I'm very confused.

I'm also wondering if I'm missing some kind of low level kernel socket setting or understanding of how this works, and (hopefully) missed something obvious that someone else will see.

I've yet to try the integration tests.

## Where to look

* net.rs - new `net` module for creating sockets with SO_REUSEADDRESS and SO_REUSEPORT
* server.rs - `spawn_downstream_receive_workers` - this is where the logic to receive packets is being managed.

## How to reproduce

Many of the tests fail in `server.rs` intermittently, but I've been debugging in`tests::run_client` - since it's the simplest test of the of the proxy with a singular packet only going from local to a single endpoint address.

To run: `cargo test --lib run_client`

## Things I have tried

My original thought was "oh, this is a race condition where we're sending the packet before the proxy comes up and is receiving connections" - there are a few ways we could tackle that, but there are several layers of async coroutines within the system, I ended up going with the tried and true approach of polling until something works. So I implemented retry logic via `tryhard` (and changing some of `test_utils` to accommodate that).

So in `run_client` I have it setup to retry 10 times with a delay of second between each attempt to sending a packet and recieve a response. Unfortunately even with the retries, it seems like intermittently, the proxy is not actively accepting packets -- and I'm not sure where it gets stuck (See logs below to see that packets don't get received by Quilkin). 

To make a note - when the test passes, it passes on the first packet that is sent. When it fails, it fails on the first packet and every retry thereafter. I've never seen a retry actually pass the test.

## Relevant logs

### Passing `run_client` unit test

You can see each of the workers come up, await for packets, accept them and process them. Working as expected (happens most of the time).

```
Testing started at 1:58 PM ...
    Finished test [unoptimized + debuginfo] target(s) in 0.11s
     Running unittests (target/debug/deps/quilkin-258efebca14997ce)
  2022-04-19T20:58:52.801777Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:49937, source: 0.0.0.0:37577
    at src/proxy/server.rs:526

  2022-04-19T20:58:52.802458Z  INFO quilkin::proxy::server: Starting, port: 49937, proxy_id: "test"
    at src/proxy/server.rs:103

  2022-04-19T20:58:52.802831Z  INFO quilkin::proxy::server: Quilkin is ready
    at src/proxy/server.rs:131

  2022-04-19T20:58:52.802906Z DEBUG quilkin::proxy::server: Awaiting packet, id: 0, addr: Ok(0.0.0.0:49937)
    at src/proxy/server.rs:228

  2022-04-19T20:58:52.802956Z DEBUG quilkin::proxy::server: Awaiting packet, id: 1, addr: Ok(0.0.0.0:49937)
    at src/proxy/server.rs:228

  2022-04-19T20:58:52.802991Z DEBUG quilkin::proxy::server: Awaiting packet, id: 2, addr: Ok(0.0.0.0:49937)
    at src/proxy/server.rs:228

  2022-04-19T20:58:52.803024Z DEBUG quilkin::proxy::server: Awaiting packet, id: 3, addr: Ok(0.0.0.0:49937)
    at src/proxy/server.rs:228

  2022-04-19T20:58:52.803057Z DEBUG quilkin::proxy::server: Awaiting packet, id: 4, addr: Ok(0.0.0.0:49937)
    at src/proxy/server.rs:228

  2022-04-19T20:58:52.803090Z DEBUG quilkin::proxy::server: Awaiting packet, id: 5, addr: Ok(0.0.0.0:49937)
    at src/proxy/server.rs:228

  2022-04-19T20:58:52.803122Z DEBUG quilkin::proxy::server: Awaiting packet, id: 6, addr: Ok(0.0.0.0:49937)
    at src/proxy/server.rs:228

  2022-04-19T20:58:52.803155Z DEBUG quilkin::proxy::server: Awaiting packet, id: 7, addr: Ok(0.0.0.0:49937)
    at src/proxy/server.rs:228

  2022-04-19T20:58:52.803305Z DEBUG quilkin::proxy::sessions::session_manager: Pruning Sessions
    at src/proxy/sessions/session_manager.rs:75

  2022-04-19T20:58:52.803385Z DEBUG quilkin::proxy::server: Received packet, id: 3, size: 5, source: 127.0.0.1:37577
    at src/proxy/server.rs:238

  2022-04-19T20:58:52.803506Z DEBUG quilkin::proxy::sessions::session: Session created, source: 127.0.0.1:37577, dest: Endpoint { address: EndpointAddress { host: Ip(0.0.0.0), port: Some(37577) }, metadata: MetadataView { known: Metadata { tokens: {} }, unknown: {} } }
    at src/proxy/sessions/session.rs:156

  2022-04-19T20:58:52.803607Z DEBUG quilkin::proxy::sessions::session: Awaiting incoming packet, source: 127.0.0.1:37577, dest: Endpoint { address: EndpointAddress { host: Ip(0.0.0.0), port: Some(37577) }, metadata: MetadataView { known: Metadata { tokens: {} }, unknown: {} } }
    at src/proxy/sessions/session.rs:181

  2022-04-19T20:58:52.803712Z DEBUG quilkin::proxy::server: Awaiting packet, id: 3, addr: Ok(0.0.0.0:49937)
    at src/proxy/server.rs:228

  2022-04-19T20:58:52.803988Z DEBUG quilkin::proxy::sessions::session: Session closed, source: 127.0.0.1:37577, dest_address: 0.0.0.0:37577
    at src/proxy/sessions/session.rs:341


Process finished with exit code 0
```

### Failing `run_client` unit test

Here you can see the socket comes up and is awaiting connections, but even with 10 retries (each "Sending Packet" debug log) with a second between, the packets are never recieved. I am very much at a loss as to why.

```

Testing started at 1:07 PM ...
    Finished test [unoptimized + debuginfo] target(s) in 0.13s
     Running unittests (target/debug/deps/quilkin-258efebca14997ce)
  2022-04-19T20:07:45.643367Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:45.644091Z  INFO quilkin::proxy::server: Starting, port: 37292, proxy_id: "test"
    at src/proxy/server.rs:103

  2022-04-19T20:07:45.644484Z  INFO quilkin::proxy::server: Quilkin is ready
    at src/proxy/server.rs:131

  2022-04-19T20:07:45.644580Z DEBUG quilkin::proxy::server: Awaiting packet, id: 0, addr: Ok(0.0.0.0:37292)
    at src/proxy/server.rs:228

  2022-04-19T20:07:45.644631Z DEBUG quilkin::proxy::server: Awaiting packet, id: 1, addr: Ok(0.0.0.0:37292)
    at src/proxy/server.rs:228

  2022-04-19T20:07:45.644667Z DEBUG quilkin::proxy::server: Awaiting packet, id: 2, addr: Ok(0.0.0.0:37292)
    at src/proxy/server.rs:228

  2022-04-19T20:07:45.644702Z DEBUG quilkin::proxy::server: Awaiting packet, id: 3, addr: Ok(0.0.0.0:37292)
    at src/proxy/server.rs:228

  2022-04-19T20:07:45.644736Z DEBUG quilkin::proxy::server: Awaiting packet, id: 4, addr: Ok(0.0.0.0:37292)
    at src/proxy/server.rs:228

  2022-04-19T20:07:45.644770Z DEBUG quilkin::proxy::server: Awaiting packet, id: 5, addr: Ok(0.0.0.0:37292)
    at src/proxy/server.rs:228

  2022-04-19T20:07:45.644804Z DEBUG quilkin::proxy::server: Awaiting packet, id: 6, addr: Ok(0.0.0.0:37292)
    at src/proxy/server.rs:228

  2022-04-19T20:07:45.644838Z DEBUG quilkin::proxy::server: Awaiting packet, id: 7, addr: Ok(0.0.0.0:37292)
    at src/proxy/server.rs:228

  2022-04-19T20:07:45.645043Z DEBUG quilkin::proxy::sessions::session_manager: Pruning Sessions
    at src/proxy/sessions/session_manager.rs:75

  2022-04-19T20:07:46.747649Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:47.849316Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:48.950998Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:50.053756Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:51.156530Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:52.258570Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:53.360224Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:54.462921Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:55.565932Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:56.669490Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:37292, source: 0.0.0.0:35604
    at src/proxy/server.rs:526

  2022-04-19T20:07:56.770633Z ERROR quilkin::proxy::health: Panic has occurred. Moving to Unhealthy, panic_info: panicked at 'called `Result::unwrap()` on an `Err` value: Elapsed(())', src/proxy/server.rs:534:10
    at src/proxy/health.rs:37


called `Result::unwrap()` on an `Err` value: Elapsed(())
thread 'proxy::server::tests::run_client' panicked at 'called `Result::unwrap()` on an `Err` value: Elapsed(())', src/proxy/server.rs:534:10
...
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: test failed, to rerun pass '-p quilkin --lib'

Process finished with exit code 101

```

If anyone has any ideas, I would love some input!

markmandel commented:
Trying something out - I extended the retry delay to 20s, and when I noticed there was a retry (and I had time) I connected using `nc`. Deliberately sending what would be a failing packet (i.e. I sent "test" doesn't match "hello", so it would fail it's `assert_equal!`) (although I sent it to `127.0.0.1` vs `0.0.0.0`) and it was received! 

Which now makes me wonder - is the issue with the sending socket, not the receiving 🤔 

Logs for postenty

```
/home/mark/.cargo/bin/cargo test --color=always --package quilkin --lib proxy::server::tests::run_client --no-fail-fast -- --format=json --exact -Z unstable-options --show-output
Testing started at 7:33 PM ...
    Finished test [unoptimized + debuginfo] target(s) in 0.14s
     Running unittests (target/debug/deps/quilkin-258efebca14997ce)
  2022-04-20T02:33:07.157913Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:51418, source: 0.0.0.0:33268
    at src/proxy/server.rs:526

  2022-04-20T02:33:07.159222Z  INFO quilkin::proxy::server: Starting, port: 51418, proxy_id: "test"
    at src/proxy/server.rs:103

  2022-04-20T02:33:07.160074Z  INFO quilkin::proxy::server: Quilkin is ready
    at src/proxy/server.rs:131

  2022-04-20T02:33:07.160284Z DEBUG quilkin::proxy::server: Awaiting packet, id: 0, addr: Ok(0.0.0.0:51418)
    at src/proxy/server.rs:228

  2022-04-20T02:33:07.160394Z DEBUG quilkin::proxy::server: Awaiting packet, id: 1, addr: Ok(0.0.0.0:51418)
    at src/proxy/server.rs:228

  2022-04-20T02:33:07.160460Z DEBUG quilkin::proxy::server: Awaiting packet, id: 2, addr: Ok(0.0.0.0:51418)
    at src/proxy/server.rs:228

  2022-04-20T02:33:07.160517Z DEBUG quilkin::proxy::server: Awaiting packet, id: 3, addr: Ok(0.0.0.0:51418)
    at src/proxy/server.rs:228

  2022-04-20T02:33:07.160577Z DEBUG quilkin::proxy::server: Awaiting packet, id: 4, addr: Ok(0.0.0.0:51418)
    at src/proxy/server.rs:228

  2022-04-20T02:33:07.160634Z DEBUG quilkin::proxy::server: Awaiting packet, id: 5, addr: Ok(0.0.0.0:51418)
    at src/proxy/server.rs:228

  2022-04-20T02:33:07.160690Z DEBUG quilkin::proxy::server: Awaiting packet, id: 6, addr: Ok(0.0.0.0:51418)
    at src/proxy/server.rs:228

  2022-04-20T02:33:07.160747Z DEBUG quilkin::proxy::server: Awaiting packet, id: 7, addr: Ok(0.0.0.0:51418)
    at src/proxy/server.rs:228

  2022-04-20T02:33:07.160986Z DEBUG quilkin::proxy::sessions::session_manager: Pruning Sessions
    at src/proxy/sessions/session_manager.rs:75

  2022-04-20T02:33:27.262431Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:51418, source: 0.0.0.0:33268
    at src/proxy/server.rs:526

  2022-04-20T02:33:36.705040Z DEBUG quilkin::proxy::server: Received packet, id: 3, size: 5, source: 127.0.0.1:52584
    at src/proxy/server.rs:238

  2022-04-20T02:33:36.705621Z DEBUG quilkin::proxy::sessions::session: Session created, source: 127.0.0.1:52584, dest: Endpoint { address: EndpointAddress { host: Ip(0.0.0.0), port: Some(33268) }, metadata: MetadataView { known: Metadata { tokens: {} }, unknown: {} } }
    at src/proxy/sessions/session.rs:156

  2022-04-20T02:33:36.706093Z DEBUG quilkin::proxy::sessions::session: Awaiting incoming packet, source: 127.0.0.1:52584, dest: Endpoint { address: EndpointAddress { host: Ip(0.0.0.0), port: Some(33268) }, metadata: MetadataView { known: Metadata { tokens: {} }, unknown: {} } }
    at src/proxy/sessions/session.rs:181

  2022-04-20T02:33:36.706553Z DEBUG quilkin::proxy::server: Awaiting packet, id: 3, addr: Ok(0.0.0.0:51418)
    at src/proxy/server.rs:228

  2022-04-20T02:33:47.364641Z DEBUG quilkin::proxy::server::tests: Sending Packet, dest: 0.0.0.0:51418, source: 0.0.0.0:33268
    at src/proxy/server.rs:526

  2022-04-20T02:33:47.364820Z ERROR quilkin::proxy::health: Panic has occurred. Moving to Unhealthy, panic_info: panicked at 'assertion failed: `(left == right)`
  left: `"hello"`,
 right: `"test\n"`', src/proxy/server.rs:537:9
    at src/proxy/health.rs:37

  2022-04-20T02:33:47.444689Z DEBUG quilkin::proxy::sessions::session: Session closed, source: 127.0.0.1:52584, dest_address: 0.0.0.0:33268
    at src/proxy/sessions/session.rs:341



<Click to see difference>

thread 'proxy::server::tests::run_client' panicked at 'assertion failed: `(left == right)`
  left: `"hello"`,
 right: `"test\n"`', src/proxy/server.rs:537:9
...
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: test failed, to rerun pass '-p quilkin --lib'

Process finished with exit code 101

```

XAMPPRocky commented:
Haven't reviewed the code yet, but I just want to mention that you might want to try debugging with [`tokio-console`](https://docs.rs/tokio-console/latest/tokio_console/) it does have some amount of helpful information.

markmandel commented:
> Haven't reviewed the code yet, but I just want to mention that you might want to try debugging with [`tokio-console`](https://docs.rs/tokio-console/latest/tokio_console/) it does have some amount of helpful information.

That's an excellent idea - I'll take that for a spin next.

XAMPPRocky commented:
Another debuggind tip that I just ran into. Setting the `span_events` for the subscriber to include entering and exiting spans as events. This is a lot noisier depending on your log level, but it's much more helpful than just events on their own.

```rust
        tracing_subscriber::fmt()
            .with_span_events(tracing_subscriber::fmt::format::FmtSpan::FULL)
            .init();
```

markmandel commented:
So unit tests are passing with that weird re-creation of a socket hack thing, but the integration tests are having the same issue and failing sometimes.

This is a very weird one.

I wrote a little bash script to run tests 100 times to make it easier to force a repro of the failure.

```shell
#!/usr/bin/env bash

set -e

for i in {1..100}; do
  echo Run "${i}"
  cargo test --test compress client_and_server
done
```

XAMPPRocky commented:
```rust
pub fn available_port() -> u16 {
     let socket = std::net::UdpSocket::bind((std::net::Ipv4Addr::UNSPECIFIED, 0)).unwrap();
     socket.local_addr().unwrap().port()
 }
```

markmandel commented:
OOOH! I have an idea. Moving to a different branch to try it out. Wish me luck.

markmandel commented:
I worked it out! OMG. That ruined me.

https://github.com/googleforgames/quilkin/blob/bac95262211165aa0788f0f5daa96e990adf774e/src/proxy/server.rs#L122

👆🏻 This line right here. I didn't think about it, but I had an extra socket I was binding that wasn't doing any `recv_from()` work, that I was using to send data back downstream.

I didn't figure that it would still receive packets, but not do anything with them. (also the whole channel/worker system for send packets downstream is no longer needed either).

Refactored it all to pass in one of the several downstream `UdpSocket` right into the `Session` so it can send data back downstream, and everything now works as expected again.

I'm going to close this PR, and do a tonne of cleanup on all my experiments, but expect some PRs and updated benchmarks etc next week.
